keps/sig-auth/3257-trust-anchor-sets/README.md

```diff
@@ -0,0 +1,1184 @@
+  well-formed ClusterTrustBundle can start up and has the expected content in
+  the targeted file.
+
+### Graduation Criteria
+
+#### Alpha
+
+In alpha, all new fields and objects are guarded by the ClusterTrustBundle
+feature gate, which is present in both kube-apiserver and kubelet.
+
+The feature will be covered by unit, integration, and E2E tests as described above.
+
+<!--
+**Note:** *Not required until targeted at a release.*
+
+Define graduation milestones.
+
+These may be defined in terms of API maturity, [feature gate] graduations, or as
+something else. The KEP should keep this high-level with a focus on what
+signals will be looked at to determine graduation.
+
+Consider the following in developing the graduation criteria for this enhancement:
+- [Maturity levels (`alpha`, `beta`, `stable`)][maturity-levels]
+- [Feature gate][feature gate] lifecycle
+- [Deprecation policy][deprecation-policy]
+
+Clearly define what graduation means by either linking to the [API doc
+definition](https://kubernetes.io/docs/concepts/overview/kubernetes-api/#api-versioning)
+or by redefining what graduation means.
+
+In general we try to use the same stages (alpha, beta, GA), regardless of how the
+functionality is accessed.
+
+[feature gate]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
+[maturity-levels]: https://git.k8s.io/community/contributors/devel/sig-architecture/api_changes.md#alpha-beta-and-stable-versions
+[deprecation-policy]: https://kubernetes.io/docs/reference/using-api/deprecation-policy/
+
+Below are some examples to consider, in addition to the aforementioned [maturity levels][maturity-levels].
+
+#### Alpha
+
+- Feature implemented behind a feature flag
+- Initial e2e tests completed and enabled
+
+#### Beta
+
+- Gather feedback from developers and surveys
+- Complete features A, B, C
+- Additional tests are in Testgrid and linked in KEP
+
+#### GA
+
+- N examples of real-world usage
+- N installs
+- More rigorous forms of testing—e.g., downgrade tests and scalability tests
+- Allowing time for feedback
+
+**Note:** Generally we also wait at least two releases between beta and
+GA/stable, because there's no opportunity for user feedback, or even bug reports,
+in back-to-back releases.
+
+**For non-optional features moving to GA, the graduation criteria must include
+[conformance tests].**
+
+[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md
+
+#### Deprecation
+
+- Announce deprecation and support policy of the existing flag
+- Two versions passed since introducing the functionality that deprecates the flag (to address version skew)
+- Address feedback on usage/changed behavior, provided on GitHub issues
+- Deprecate the flag
+-->
+
+### Upgrade / Downgrade Strategy
+
+At present, there are no upgrade / downgrade concerns.  The ClusterTrustBundle feature gate controls overall availability of the feature.
```

Comment:
I think that admissionwebhook configuration and apiservice behavior with empty cabundles represents a migration concern on downgrade.

Reply:
Suggestion: "The addition of ClusterTrustBundles offers the opportunity for teams to migrate Webhook Configurations from directly embedded CA Bundles ({mutatingwebhookconfigurations,validatingwebhookconfigurations}.webhooks.clientConfig.caBundle) to external ClusterTrustBundle references. Administrators making use of the new {mutatingwebhookconfigurations,validatingwebhookconfigurations}.webhooks.clientConfig.trustAnchors.clusterTrustBundleName field would need to revert back to using the caBundle field before/immediately after downgrading, particularly when `failurePolicy = Fail`"

Reply:
I think this is now handled by saying that both CABundle and ClusterTrustBundleName can be set at the same time, but that ClusterTrustBundleName takes precedence.